### PR TITLE
Add Bazel tooling support

### PR DESCRIPTION
**Pull request by cassava**
You still need to supply your own `.bazelrc.user` file with a registry containing any missing packages. For example:

```
common --registry=file:///home/youser/my-bazel-registry
```

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# bazel from apt needs access to this cacerts location
+startup --host_jvm_args=-Djavax.net.ssl.trustStore=/etc/ssl/certs/java/cacerts
+
+# So we can support private dependencies
+build --experimental_cc_implementation_deps
+
+build --action_env=BAZEL_CXXOPTS="-std=c++17"
+build --strip=never
+build --copt='-O2'
+build --conlyopt='-std=c99'
```
Comment (BenjaminKern) [inline .bazelrc:14]: i think we could go back to c11 (but c99 is fine i guess)

File: .gitignore [hunk 2]
```diff
@@ -58,3 +58,16 @@ callgrind.out.*
 # Miscellaneous
 *.log
 /.gtm/
+
+# Ignore all bazel-* symlinks. There is no full list since this can change
+# based on the name of the directory bazel is cloned into.
+/bazel-*
+MODULE.bazel.lock
+.bazelrc.user
+
```
Comment (BenjaminKern) [inline .gitignore:67]: hedron compile commands adds an external folder which you can also ignore if you want

File: BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,8 @@
+load("//:plugin.bzl", "CLOE_PLUGINS")
```
Comment (BenjaminKern) [inline BUILD.bazel:1]: not sure if i would keep the *.bzl on the top lvl (i have a feeling that in the long run there will be more *.bzl files in the long run, which might either lead to a cluttered top lvl or all over the repo distributes bzl files which is imho not so pleasant)

File: MODULE.bazel [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+# Build dependencies:
+bazel_dep(name = "cmake_configure_file", version = "0.1.0")
+bazel_dep(name = "rules_cc", version = "0.0.9")
+
+# Test dependencies:
+bazel_dep(name = "googletest", version = "1.14.0", repo_name = "gtest")
+
+# Library dependencies:
+bazel_dep(name = "boost", version = "1.83.0.bcr.1")
```
Comment (BenjaminKern) [inline MODULE.bazel:11]: provide a link to the PR where this boost module is proposed to added to the bcr

Comment (BenjaminKern) [reply MODULE.bazel:11]: Also mark all the modules that are not officially available in the bcr yet, so a user knows for which libraries he has to manually provide a bazel module

File: README.md [hunk 5]
```diff
@@ -194,6 +194,35 @@ When using `make` to build the project, add this to the command line:
...
+This file is ignored by Git to prevent you from exposing your secrets.
+The registry should contain the following modules:
+
+    boost
+    cli11
+    incbin
+    inja
+    oatpp
+    open-simulation-interface
+    sol
```
Comment (BenjaminKern) [inline README.md:222]: ah nevermind my ohter comment but maybe add the fact that apart from boost those other modules can be provided easily and link ecplicitly to the boost pr for the bcr

File: cloe_shell.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (BenjaminKern) [inline cloe_shell.sh:1]: consider using the shebang /usr/bin/env bash

File: cloe_shell.sh [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+export CLOE_LOG_LEVEL=debug
+export CLOE_STRICT_MODE=1
+export CLOE_ROOT="$(pwd)"
```
Comment (BenjaminKern) [inline cloe_shell.sh:5]: hmm is pwd correct here or does it really need to be the root of this repository? If yes maybe consider using basename of this script shenanigans

File: engine/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,66 @@
+load("//:version.bzl", "PROJECT_VERSION")
```
Comment (BenjaminKern) [inline engine/BUILD.bazel:1]: ditto about *.bzl. (again just optics not really necessary to do any actions here)

File: engine/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+cc_library(
```
Comment (BenjaminKern) [inline engine/BUILD.bazel:3]: if you name this library engine you can consume it via //engine.  If not you have to use the following //engine:enginelib

File: engine/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            "src/**/*_test.cpp",
+            "src/server_mock.cpp",
+        ]
+    ),
+    data = glob(["lua/**"]),
+    includes = ["src"],
+    additional_compiler_inputs = glob(["webui/**"]),
+    deps = [
+        "//engine/vendor/lrdb",
+        "//engine/vendor/linenoise",
```
Comment (BenjaminKern) [inline engine/BUILD.bazel:19]: I know that is not possible to that cleanly, while also supporting cmake but consider consuming all vendor files (linenoise and so on) as bazel modules in the long run (you then just have cleanly documented all third-party dependencies that way as well)

File: engine/BUILD.bazel [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        "//stack",
+        "//fable",
+        "//runtime",
+        "//models",
+        "//oak",
+        "@boost//:algorithm",
+        "@boost//:conversion",
+        "@boost//:filesystem",
```
Comment (BenjaminKern) [inline engine/BUILD.bazel:27]: nitpick, hopefully you can drop this dependency eventually when you drop < gcc-9.2 support

File: oak/BUILD.bazel [hunk 12]
```diff
@@ -0,0 +1,26 @@
+cc_library(
+    name = "oak",
+    hdrs = glob(["include/**/*.hpp"]),
+    srcs = glob(["src/**"], exclude=["src/**/*_test.cpp"]),
+    includes = [ "include" ],
+    copts = [
+        "-Ioak/src"
```
Comment (BenjaminKern) [inline oak/BUILD.bazel:7]: why is this necessary?

File: .bazelrc (hunk 1 above)
Comment (cassava) [reply .bazelrc:14]: No you're right. C11 has better support than C99.

File: .gitignore (hunk 2 above)
Comment (cassava) [reply .gitignore:67]: Good to know! Do you know which one?

File: BUILD.bazel (hunk 3 above)
Comment (cassava) [reply BUILD.bazel:1]: Do you think it's acceptable to reorganize this for making it leave the experimental stage?

File: cloe_shell.sh (hunk 6 above)
Comment (cassava) [reply cloe_shell.sh:1]: I see this for python but have rarely seen it for Bash. Is it important for Bazel specifically?

There is: https://unix.stackexchange.com/questions/206350/what-is-the-difference-if-i-start-bash-with-bin-bash-or-usr-bin-env-bash

File: cloe_shell.sh (hunk 7 above)
Comment (cassava) [reply cloe_shell.sh:5]: It is correct here because cloe_shell is started by BUILD.bazel target which has the root dir. It's just defined for compatibility with some of the tests that want to know.

File: engine/BUILD.bazel (hunk 9 above)
Comment (cassava) [reply engine/BUILD.bazel:3]: I know. I do not want it to be used outside just yet. If anything, I'd want `cloe-engine` binary target to be named just `engine`, so that it can be run more easily. Until we decide, better to not have a default yet.

File: engine/BUILD.bazel (hunk 10 above)
Comment (cassava) [reply engine/BUILD.bazel:19]: Agreed, but that would require more changes to all the CMake/Conan stuff, and I want to add experimental Bazel tooling that works with the sources as-is.

File: oak/BUILD.bazel (hunk 12 above)
Comment (cassava) [reply oak/BUILD.bazel:7]: It's probably not...

Comment (cassava) [reply oak/BUILD.bazel:7]: How would you define *local includes*. There are headers not part of the public interface that I want to include and they are in the `src` folder.

File: engine/BUILD.bazel (hunk 11 above)
Comment (cassava) [reply engine/BUILD.bazel:27]: Can already be removed, actually.

File: engine/BUILD.bazel (hunk 8 above)
Comment (cassava) [reply engine/BUILD.bazel:1]: I want the place where version of the whole project is defined to be at repository root. It needs to be changed often, it's important, and it should be obvious. I don't want it languishing in `tools/bazel/version.bzl` where one would expect... helper tools, not important project configuration.

Perhaps I should rename it to `config.bzl` or something.

For the other one, `plugin.bzl` I guess that could go somewhere else.

File: README.md (hunk 5 above)
Comment (cassava) [reply README.md:222]: Bazel support is experimental and requires the user to know what they are doing. If someone already has a Bazel infrastructure I trust they will find the dependencies.

The other modules can be provided easily is somewhat of an understatement if you are not already using Bazel. And if you are, then you likely already know. But maybe I put this in the readme.

File: BUILD.bazel (hunk 3 above)
Comment (cassava) [reply BUILD.bazel:1]: That is: I want to do it later, when Bazel is more officially supported, and is not experimental.

File: .bazelrc [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+common --registry=https://bcr.bazel.build
+try-import %workspace%/.bazelrc.user
+
+# bazel from apt needs access to this cacerts location
+startup --host_jvm_args=-Djavax.net.ssl.trustStore=/etc/ssl/certs/java/cacerts
+
+# So we can support private dependencies
+build --experimental_cc_implementation_deps
+
+build --action_env=BAZEL_CXXOPTS="-std=c++17"
```
[the remark this answers is not in the document]
Comment (cassava) [reply .bazelrc:11]: Yes, that would be too much work for now.

File: MODULE.bazel (hunk 4 above)
Comment (cassava) [reply MODULE.bazel:11]: They are marked in the README, and I don't really want to provide a link to a PR for Boost support, because I do not want to take responsibility for that.

File: README.md (hunk 5 above)
Comment (cassava) [reply README.md:222]: It's already in the README.

Comment (cassava) [reply README.md:222]: (See my other comment.)

File: oak/BUILD.bazel (hunk 12 above)
Comment (cassava) [reply oak/BUILD.bazel:7]: https://github.com/bazelbuild/bazel/issues/15553#issuecomment-1134954324

So apparently this is the only way right now. A bit disappointing.

Comment (BenjaminKern) [reply oak/BUILD.bazel:7]: just add the private headers to the srcs glob as well (we are doing this as well in various places).

File: .gitignore (hunk 2 above)
Comment (BenjaminKern) [reply .gitignore:67]: the folder is called „external“ 😆

File: README.md (hunk 5 above)
Comment (BenjaminKern) [reply README.md:224]: Sure I am implying a somewhat familiarity with bazel here

File: MODULE.bazel (hunk 4 above)
Comment (cassava) [reply MODULE.bazel:11]: Exactly :-D

File: oak/BUILD.bazel (hunk 12 above)
Comment (cassava) [reply oak/BUILD.bazel:7]: Ok, it works if I include relative to the source file.

File: MODULE.bazel (hunk 4 above)
Comment (cassava) [reply MODULE.bazel:11]: Hey look here: https://github.com/bazelboost/registry

File: .gitignore (hunk 2 above)
Comment (cassava) [reply .gitignore:68]: Thanks!
